trackintel/preprocessing/triplegs.py

```diff
@@ -89,283 +86,204 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, print_progress=Fals
+    spts_tpls["first_activity"] = first_activities
+    spts_tpls["last_activity"] = last_activities  # trip may start here
+    spts_tpls = spts_tpls[~inter_activities]  # no trip interaction
+    spts_tpls["started_at_next"] = spts_tpls["started_at"].shift(-1)
+
+    # conditions for new trip
+    # start new trip if the user changes
+    condition_new_user = spts_tpls["user_id"] != spts_tpls["user_id"].shift(1)
+
+    # start new trip if there is a new activity
+    condition_new_activity = spts_tpls["last_activity"]
+
+    # gap conditions
+    # start new trip after a gap
+    gap = (spts_tpls["started_at_next"] - spts_tpls["finished_at"]) > gap_threshold
+    condition_time_gap = gap.shift(1, fill_value=False)  # trip starts on next entry
+
+    cond_all = condition_new_user | condition_new_activity | condition_time_gap
+    spts_tpls["new_trip"] = cond_all
+
+    # assign an incrementing id to all triplegs that start a trip
+    spts_tpls.loc[cond_all, "trip_id"] = np.arange(cond_all.sum())
+    spts_tpls["trip_id"].fillna(method="ffill", inplace=True)
+
+    # exclude activities to aggregate trips together.
+    # activity can be thought of as the same level as trips.
+    spts_tpls_no_act = spts_tpls[~spts_tpls["activity"]]
+    spts_tpls_only_act = spts_tpls[spts_tpls["activity"]]
+
+    trips_grouper = spts_tpls_no_act.groupby("trip_id")
+    trips = trips_grouper.agg({"user_id": "mean", "started_at": min, "finished_at": max, "type": list, "id": list})
+
+    def _seperate_ids(row):
+        """Split aggregated ids into staypoints and triplegs arrays."""
+        row_type = np.array(row["type"])
+        row_id = np.array(row["id"])
+        t = row_type == "tripleg"
+        tpls_ids = row_id[t]
+        spts_ids = row_id[np.logical_not(t)]
+        return [spts_ids, tpls_ids]
+
+    trips[["spts", "tpls"]] = trips.apply(_seperate_ids, axis=1, result_type="expand")
+
+    # recount trips ignoring empty trips and save trip_id as for id assignment.
+    trips.reset_index(inplace=True)
+    trips["trip_id"] = trips.index
+
+    # add gaps as activities, to simplify id assignment.
+    gaps = pd.DataFrame(spts_tpls.loc[gap, "user_id"])
+    gaps["started_at"] = spts_tpls.loc[gap, "finished_at"] + gap_threshold / 2
+    gaps[["type", "activity", "new_trip"]] = ["gap", True, True]
+
+    # same for user changes
+    user_change = pd.DataFrame(spts_tpls.loc[condition_new_user, "user_id"])
+    user_change["started_at"] = spts_tpls.loc[condition_new_user, "started_at"] - gap_threshold / 2
+    user_change[["type", "activity", "new_trip"]] = ["user_change", True, True]
+
+    # merge with activities
+    trips.drop(columns=["type", "id"], inplace=True)  # make space so no overlap with activity "id"
+    spts_tpls_only_act.drop(columns=["trip_id"])  # no overlap of activity trip_ids to real trip_ids in trips
+    trips = pd.concat((trips, spts_tpls_only_act, gaps, user_change), axis=0, ignore_index=True)
```

Comment:
I am confused by the `type`, `id` and `trip_id` columns for each of the three different df at this point. Is there a way to increase the readability here?

Also, the `trips` here is not yet really trips? In line 204 we get the staypoints from "trips". I guess here we are working on activitySpts_trips?

Reply:
I renamed `"id"` to `"spts_tpls_id"` and the temporary trip_ids to `"temp_trip_id"` and the final trip_id remains `"trip_id"`. I left `"type"` as the name because I reuse that for gaps, new_user activities and find the name appropriate and not so confusing when you change the others.

Yes, these are really trips with activities, I renamed it accordingly to show the differences.
I think with these changes the function will be much cleaner. Thanks!